internal/beater/api/mux.go

```diff
@@ -254,16 +254,7 @@ func agentConfigHandler(
 	fleetManaged bool,
 ) (request.Handler, error) {
 	mw := middlewareFunc(cfg, authenticator, ratelimitStore, agent.MonitoringMap)
-	h := agent.NewHandler(f, cfg.KibanaAgentConfig.Cache.Expiration, cfg.DefaultServiceEnvironment, cfg.AgentAuth.Anonymous.AllowAgent)
-
-	if !cfg.Kibana.Enabled && !fleetManaged {
-		msg := "Agent remote configuration is disabled. " +
```

Comment:
What behavior do we want here? Do we want to maintain the 403 with error message or just log it?

Reply:
Alternatively we can expose a Blocked function in Fetcher interface to decouple the implementation and facilitate testing but would love to know your thoughts.

Reply:
> It is not impossible to maintain the 403 but it is increasingly tricky to test this because ElasticsearchFetcher will block Fetch requests until background job finishes for the first time. And the mux tests are definitely not designed to know that much implementation details.

Also, just because it failed the first time doesn't mean it will fail forever. So I don't think it makes sense to log this message up-front when using the Elasticsearch output. Should we still log the error when some other output is configured?

This is the kind of behaviour I'd expect:
 1. If `apm-server.agent_config` is defined, use that.
 1. If `output.elasticsearch` is not configured, and `apm-server.kibana` is not configured, log the message and respond 403 early. Otherwise:
 1. If `output.elasticsearch` is configured, combine it with `apm-server.agent.config.elasticsearch.*` and attempt to query Elasticsearch for config. If that succeeds, stop.
 1. If `apm-server.kibana` is configured, use the old Kibana method.
 1. If each of the above methods failed (due to an error, not just because there's no matching config), log a message and respond 403.

Reply:
My current implementation is slightly different from what you just mentioned. The difference is priority of `apm-server.agent_config` vs ES Fetcher.

In the background:
1. It will always query ES periodically using a mix of `default ES config + output.elasticsearch + apm-server.agent.config.elasticsearch.*`.
2. If the ES query returns HTTP 200, it will flag using `expectedIndexExists` that no fallback should be used.

On every fetch:
1. Fetch request will be blocked until the first background job completes.
2. If the `expectedIndexExists` flag is true, fetch from the cache populated by background job.
3. If the `expectedIndexExists` flag is false, fetch using the fallback method (either through `apm-server.agent_config` or send query to kibana on demand.

I guess the easiest way is if `expectedIndexExists` is false, and there is no kibana and agent_config configured, return 403.

Reply:
Updated implementation will return 403 when ElasticsearchFetcher fails without a fallback. This behavior is tested using system test instead of a mux middleware unit test.

Reply:
>Fetch request will be blocked until the first background job completes.

This is problematic if ES e.g. is unreachable at startup. Instead of blocking until the apm agent finally closes the connection or a connection to ES can be established at some point, I'd expect a very short timeout server side, after which the server applies the regular fallback logic.

Reply:
I agree that this can be problematic. Also, when we say timeout, I am thinking about the timeout for the entire refreshCache operation to complete, i.e. initial search request + additional pagination requests. Can we safely assume that it will all finish in e.g. 5 seconds?

Reply:
I'd be very worried if this operation takes longer than 5 seconds and we are locking the cache for this whole time. 

As the cache is updated async, I don't think that we should be blocking the incoming request in case of an ongoing update. The default response signals the agents to query again in 30 seconds. So if the request comes in while the underlying cache is updated, IMO it is fine to still return the previous configuration, we can accept a delay of 30seconds for any new config values. 

For the startup scenario, when the underlying ES is not yet available, we dont know if the fallback or the new logic should be applied, and if ES is not available, then most certainly Kibana would also not yet be available. So one option would be to return a `503` for the time being, until everything is started up and reachable.

Reply:
It will only be blocked during startup when there is no cache to serve incoming requests. Subsequent requests are always served using the cache immediately.

But I think that 503 is a better solution as long as it does not cause problems to the clients and that they will retry after a reasonable time.

Reply:
@axw first suggested blocking on a channel (and ctx) until the 1st refresh cache attempt has been made. What do you think about returning 503 vs blocking incoming requests on startup when the cache is not initialized?

Reply:
Agreed, responding 503 sounds like a good idea.

Reply:
@carsonip please also check the agent spec for the agent config implementation to ensure that this is covered and all agents can handle it (assume they can, but let's double check).

Reply:
Returning a 503 when cache is cold now. This behavior will exist for at max 5 seconds from startup. After that, it will either serve fetch requests using the cache (if the refresh cache was successful) or fallback (if refresh cache timed out or failed due to other reasons)

Reply:
https://www.elastic.co/guide/en/apm/guide/current/api-config.html
In the spec, 503 is for kibana is unreachable, so I assume it is handled by agents. We do have to update the docs about the 403 and 503 to reflect the changes that:
403: Elasticsearch does not work and fallback to kibana fails
503: Server is starting up or Kibana is unreachable in the case of fallback

Reply:
done

Reply:
As suggested by @simitt, there will be no more 403 and it will always use 503 for both startup behavior and in the case of unreachable ES + no fallback.

Reply:
Apologies for the long back and forth on this @carsonip. You changed to always return a `503` as I suggested. My intention was to streamline behavior and make it easier to follow what is really going on. 
I tested your current implementation, and a couple of things are still confusing to me when running them. Trying to summarize what I'd expect to see, but if you're up for it, might be eaiser to spend an hour together to have a live conversation about it while running the code.  

- If `apm-server.agent_config.elasticsearch` is defined, then it should be used, and agentcfg be queried from ES, no fallback. If the request is not succesful, log the reason (e.g. unauthorized or ES not reachable). The config option `apm-server.agent_config.elasticsearch` is newly introduced, so it indicates that the apm-server should use the new implementation. Am I missing any reason why we should apply a fallback for this case or return the previously used `403` with the log message to configure the kibana section?
- Otherwise, if `kibana` is configured, apply the fallback to fetch from Kibana.
- Otherwise, fail fast and return that the configuration for agentcfg is missing.